src/core/unit.c

```diff
@@ -1535,56 +1543,84 @@ static int unit_add_slice_dependencies(Unit *u) {
-                        if (r < 0)
-                                return r;
+                m = manager_get_unit(u->manager, p);
+                if (!m) {
+                        /* Make sure to load the mount unit if it exists. If so the dependencies on
+                         * this unit will be added later during the loading of the mount unit. */
+                        (void) manager_load_unit_prepare(u->manager, p, /* path= */NULL, /* e= */NULL, &m);
+                        continue;
+                }
+                if (m == u)
+                        continue;
 
-                        m = manager_get_unit(u->manager, p);
-                        if (!m) {
-                                /* Make sure to load the mount unit if it exists. If so the dependencies on
-                                 * this unit will be added later during the loading of the mount unit. */
-                                (void) manager_load_unit_prepare(u->manager, p, NULL, NULL, &m);
-                                continue;
-                        }
-                        if (m == u)
-                                continue;
+                if (m->load_state != UNIT_LOADED)
+                        continue;
 
-                        if (m->load_state != UNIT_LOADED)
-                                continue;
+                r = unit_add_dependency(u, UNIT_AFTER, m, /* add_reference= */ true, di.origin_mask);
+                if (r < 0)
+                        return r;
+                changed = changed || r > 0;
 
-                        r = unit_add_dependency(u, UNIT_AFTER, m, true, di.origin_mask);
+                if (m->fragment_path) {
+                        r = unit_add_dependency(u, d, m, /* add_reference= */ true, di.origin_mask);
                         if (r < 0)
                                 return r;
                         changed = changed || r > 0;
-
-                        if (m->fragment_path) {
-                                r = unit_add_dependency(u, UNIT_REQUIRES, m, true, di.origin_mask);
-                                if (r < 0)
-                                        return r;
-                                changed = changed || r > 0;
-                        }
                 }
         }
 
+        *ret_changed = changed || *ret_changed;
+
+        return 0;
+}
+
+static int unit_add_mount_dependencies(Unit *u) {
+        UnitDependencyInfo di;
+        const char *path;
+        bool changed = false;
+        int r;
+
+        assert(u);
+
+        HASHMAP_FOREACH_KEY(di.data, path, u->requires_mounts_for) {
+                r = add_mount_one(u, di, path, UNIT_REQUIRES, &changed);
+                if (r < 0)
+                        return r;
+        }
+
+        HASHMAP_FOREACH_KEY(di.data, path, u->wants_mounts_for) {
+                r = add_mount_one(u, di, path, UNIT_WANTS, &changed);
+                if (r < 0)
+                        return r;
+        }
```

Comment:
I really don't like this duplication for all these functions. Can we maybe add `UnitMountDependencyType` as a new enum with UNIT_MOUNT_REQUIRES and UNIT_MOUNT_WANTS.

And then always loop over 0…UNIT_MOUNT_DEPENDENCY_TYPE_MAX, and so on?

You get the idea

Reply:
I don't follow, I've already split out the common helpers?